provisioning_client/adapters/hsm_client_x509.c

```diff
@@ -0,0 +1,196 @@
+            result = MU_FAILURE;
+        }
+        else if (mallocAndStrcpy_s(&x509_client->x509certificate, certificate) != 0)
+        {
+            LogError("Failed allocating certificate key");
+            result = MU_FAILURE;
+        }
+        else
+        {
+            result = 0;
+        }
+    }
+    return result;
+}
+
+int hsm_client_x509_set_key(HSM_CLIENT_HANDLE handle, const char* key)
+{
+    int result;
+    if (handle == NULL || key == NULL)
+    {
+        LogError("Invalid parameter handle: %p, key: %p", handle, key);
+        result = MU_FAILURE;
+    }
+    else
+    {
+        HSM_CLIENT_X509_INFO* x509_client = (HSM_CLIENT_X509_INFO*)handle;
+        if (x509_client->x509key != NULL)
+        {
+            LogError("Certificate key has been previously set and cannot be changed");
+            result = MU_FAILURE;
+        }
+        else if (mallocAndStrcpy_s(&x509_client->x509key, key) != 0)
+        {
+            LogError("Failed allocating key key");
+            result = MU_FAILURE;
+        }
+        else
+        {
+            result = 0;
+        }
+    }
+    return result;
+}
+
+char* hsm_client_x509_get_certificate(HSM_CLIENT_HANDLE handle)
+{
+    char* result;
+    if (handle == NULL)
+    {
+        LogError("Invalid handle value specified.");
+        result = NULL;
+    }
+    else
+    {
+        HSM_CLIENT_X509_INFO* x509_client = (HSM_CLIENT_X509_INFO*)handle;
+
+        if (x509_client->x509certificate == NULL)
+        {
+            LogError("Client X509 certificate was not configured.");
+            result = NULL;
+        }
+        else
+        {
+            // Create a copy. Lifetime is managed by the DPS or Hub clients.
+            if (mallocAndStrcpy_s(&result, x509_client->x509certificate) != 0)
+            {
+                LogError("Failure allocating x509certificate.");
+                result = NULL;
+            }
+        }
+    }
+
+    return result;
+}
+
+char* hsm_client_x509_get_key(HSM_CLIENT_HANDLE handle)
+{
+    char* result;
+    if (handle == NULL)
+    {
+        LogError("Invalid handle value specified.");
+        result = NULL;
+    }
+    else
+    {
+        HSM_CLIENT_X509_INFO* x509_client = (HSM_CLIENT_X509_INFO*)handle;
+
+        if (x509_client->x509key == NULL)
+        {
+            LogError("Client X509 key was not configured.");
+            result = NULL;
+        }
+        else
+        {
+            // Create a copy. Lifetime is managed by the DPS or Hub clients.
+            if (mallocAndStrcpy_s(&result, x509_client->x509key) != 0)
+            {
+                LogError("Failure allocating x509key.");
+                result = NULL;
+            }
+        }
+    }
+    
+    return result;
+}
+
+char* hsm_client_x509_get_common_name(HSM_CLIENT_HANDLE handle)
+{
+    (void)handle;
+    // Provisioning Device Client should never call this function.
+    LogError("Registration ID was not configured.");
```

Comment:
This is one of the "hacks" I had to do to avoid API changes. The way this works is that the application must call
```C
Prov_Device_LL_SetOption(user_ctx.handle, PROV_REGISTRATION_ID, registration_id)
```

If that is set, it will override the HSM interface implementation, and the code will never get to call `get_common_name`. 
If the application did not use the option, then this HSM interface function gets called and we don't have the information here. 

The idea behind the HSM interface `get_common_name` is that a HSM may return a certificate and a key (or key handle) and the developer must supply the registration ID. Most likely this will be a constant or, it can be obtained by parsing the X509 Certificate's Subject field.